source/nrnmatlab.cpp

```diff
@@ -271,4 +272,139 @@ void NrnRef::set(double x) {
 }
 double NrnRef::get() {
     return *(this->ref);
-}
\ No newline at end of file
+}
+
+std::vector<double> get_x3d(Section* sec) {
+    auto result = std::vector<double>(sec->npt3d);
+    for (size_t i = 0; i < sec->npt3d; i++) {
+        result[i] = sec->pt3d[i].x;
+    }
+    return result;
+}
+
+std::vector<double> get_y3d(Section* sec) {
+    auto result = std::vector<double>(sec->npt3d);
+    for (size_t i = 0; i < sec->npt3d; i++) {
+        result[i] = sec->pt3d[i].y;
+    }
+    return result;
+}
+
+std::vector<double> get_z3d(Section* sec) {
+    auto result = std::vector<double>(sec->npt3d);
+    for (size_t i = 0; i < sec->npt3d; i++) {
+        result[i] = sec->pt3d[i].z;
+    }
+    return result;
+}
+
+std::vector<double> get_arc3d(Section* sec) {
+    auto result = std::vector<double>(sec->npt3d);
+    for (size_t i = 0; i < sec->npt3d; i++) {
+        result[i] = sec->pt3d[i].arc;
+    }
+    return result;
+}
+
+std::vector<double> get_d3d(Section* sec) {
+    auto result = std::vector<double>(sec->npt3d);
+    for (size_t i = 0; i < sec->npt3d; i++) {
+        result[i] = sec->pt3d[i].d;
+    }
+    return result;
+}
+
+// Linearly interpolates between a and b, extrapolates if f is outside the range [0, 1].
+// Added since we don't use c++ 20 or above where we could use:
+// https://en.cppreference.com/w/cpp/numeric/lerp
+double linearly_interpolate(double a, double b, double f) {
```

Comment:
Can we add a small test with 3d points on a section to check if all this interpolation functionality returns the right values?

Reply:
I'm not sure where `addpoint` was added (probably not this PR), but... can we rename to `pt3dadd` to be consistent with Python?

Reply:
```matlab
main.push();
n.pt3dadd(201, 2, 3, 1);
n.pop_section();
```
Addpoint is our own function which circumvents the need for pushing and popping of sections

Reply:
Just to point this out... `.length` should not be needed here, and it *should* be overriden by the shape defined by those 4 points (a section passing through them would, I believe, have a length bigger than 10).

Reply:
Yes, indeed, you can add a section as the first function argument and it will push it onto the session stack, and pop it off after the function is done.

I think it's fine to rename `Section.addpoint` to `Section.pd3dadd`.